File: ui/src/api/projects.ts
/**
 * Projects API
 */

import { invoke } from '@tauri-apps/api/core';
import type { CommandResponse, Project, ProjectStatus } from '../types';

export async function listProjects(): Promise<Project[]> {
  const response = await invoke<CommandResponse<{ projects: Project[]; total: number }>>(
    'list_projects'
  );
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to list projects');
  }
  return response.data.projects;
}

export async function addProject(path: string): Promise<Project> {
  const response = await invoke<CommandResponse<{ project: Project }>>('add_project', {
    path,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to add project');
  }
  return response.data.project;
}

export async function removeProject(path: string): Promise<void> {
  const response = await invoke<CommandResponse<unknown>>('remove_project', {
    path,
  });
  if (!response.success) {
    throw new Error(response.error || 'Failed to remove project');
  }
}

export async function getProjectStatus(path: string): Promise<ProjectStatus> {
  const response = await invoke<CommandResponse<ProjectStatus>>('get_project_status', {
    path,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get project status');
  }
  return response.data;
}

export async function openProjectFolder(path: string): Promise<void> {
  const response = await invoke<CommandResponse<void>>('open_project_folder', {
    path,
  });
  if (!response.success) {
    throw new Error(response.error || 'Failed to open project folder');
  }
}

export async function initProject(
  path: string,
  preset?: string
): Promise<{ path: string; preset?: string }> {
  const response = await invoke<CommandResponse<{ path: string; preset?: string }>>(
    'init_project',
    { path, preset }
  );
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to initialize project');
  }
  return response.data;
}

File: ui/src-tauri/src/commands/projects.rs
use super::{bridge_call, CommandResponse};
use crate::bridge::BridgeManager;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;
use tauri::State;

#[derive(Debug, Serialize, Deserialize)]
pub struct Project {
    pub name: String,
    pub path: String,
    pub configured_at: Option<String>,
    pub last_accessed: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProjectStatus {
    pub name: String,
    pub path: String,
    pub has_devflow_config: bool,
    pub infrastructure_enabled: bool,
    pub services_running: u32,
    pub services_total: u32,
}

/// List all registered projects
#[tauri::command]
pub fn list_projects(bridge: State<Arc<BridgeManager>>) -> CommandResponse<Value> {
    match bridge_call(&bridge, "projects.list", None) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Add a project to the registry
#[tauri::command]
pub fn add_project(bridge: State<Arc<BridgeManager>>, path: String) -> CommandResponse<Value> {
    match bridge_call(&bridge, "projects.add", Some(json!({ "path": path }))) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Remove a project from the registry
#[tauri::command]
pub fn remove_project(bridge: State<Arc<BridgeManager>>, path: String) -> CommandResponse<Value> {
    match bridge_call(&bridge, "projects.remove", Some(json!({ "path": path }))) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Get project status
#[tauri::command]
pub fn get_project_status(
    bridge: State<Arc<BridgeManager>>,
    path: String,
) -> CommandResponse<Value> {
    match bridge_call(&bridge, "projects.status", Some(json!({ "path": path }))) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Open project in file manager
#[tauri::command]
pub fn open_project_folder(path: String) -> CommandResponse<()> {
    match open::that(&path) {
        Ok(()) => CommandResponse::ok(()),
        Err(e) => CommandResponse::err(format!("Failed to open folder: {}", e)),
    }
}

/// Initialize a new project with devflow.yml
#[tauri::command]
pub fn init_project(
    bridge: State<Arc<BridgeManager>>,
    path: String,
    preset: Option<String>,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "projects.init",
        Some(json!({ "path": path, "preset": preset })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}
